### ISSUE TITLE
Switching project broken after update.

### ISSUE THREAD

**Issue by thblt** [ericdanan/counsel-projectile#19]
After a recent update, switching project fails with either "You're not in a project" or a buffer being created for a non-existing file.

## Steps to reproduce

From `emacs -Q`, eval:

``` elisp
(require 'package)
(package-initialize)

(require 'projectile)
(require 'counsel-projectile)
(projectile-global-mode)
(counsel-projectile-on)
```

`C-c p p`, pick a project, pick a file.
Result: 
 - *If the current buffer (the one in which you typed `C-c p p`) is in a project*: a new buffer gets created for a non existing path in current project, but with the path and name of the file you've tried to open.  Eg, if you try the step above from `~/Code/ProjectA/somefile.c` and try to open `~/Code/ProjectB/src/SomeOtherFile.hs`, the new buffer is for the file `~/Code/ProjectA/src/SomeOtherFile.hs`.  If the base path doesn't exist, you get the usual `type M-x make-directory etc` message.
 - *Otherwise*: an error message "You're not in a project" appears and nothing happens. 

## Attempted solutions

I tried to

 - delete Projectile cache and bookmarks.
 - delete `.emacs.d/elpa` and reinstall everything.

Without success.

## Configuration

Debian Stretch x86_64, Emacs 24.5.1 from official repositories, all packages from Melpa.

The bug **cannot** be reproduced in the same machine with Emacs 25 from official repositories, all packages (by order of priority) from: Melpa Stable, then Gnu, then Melpa.

**Comment by thblt** [comment]
I forgot the debugger backtrace (after `toggle-debug-on-error`)

```
Debugger entered--Lisp error: (error "You're not in a project")
  signal(error ("You're not in a project"))
  error("You're not in a project")
  projectile-project-root()
  projectile-expand-root(#(".emacs.d/init-minimal.el" 0 1 (type file face ivy-virtual) 1 24 (face ivy-virtual)))
  counsel-projectile--find-file-action(#(".emacs.d/init-minimal.el" 0 1 (type file face ivy-virtual) 1 24 (face ivy-virtual)) nil)
  counsel-projectile--action(#(".emacs.d/init-minimal.el" 0 1 (type file face ivy-virtual) 1 24 (face ivy-virtual)))
  ivy-call()
  ivy-read("[.dotfiles] Load buffer or file: " (#(".emacs.d/init-minimal.el" 0 1 (face ivy-virtual type file) 1 24 (face ivy-virtual)) #(".emacs.d/user-lisp/#setup-projectile.el#" 0 1 (face ivy-virtual type file) 1 40 (face ivy-virtual)) #(".emacs.d/user-lisp/setup-org.el" 0 1 (face ivy-virtual type file) 1 31 (face ivy-virtual)) #(".Xmodmap" 0 1 (face ivy-virtual type file) 1 8 (face ivy-virtual)) #(".Xresources" 0 1 (face ivy-virtual type file) 1 11 (face ivy-virtual)) #(".bin/anybrightness" 0 1 (face ivy-virtual type file) 1 18 (face ivy-virtual)) #(".bin/apt-what-have-I-installed" 0 1 (face ivy-virtual type file) 1 30 (face ivy-virtual)) #(".bin/bd" 0 1 (face ivy-virtual type file) 1 7 (face ivy-virtual)) #(".bin/bdsk2bib" 0 1 (face ivy-virtual type file) 1 13 (face ivy-virtual)) #(".bin/bootstrap_osx" 0 1 (face ivy-virtual type file) 1 18 (face ivy-virtual)) #(".bin/debian_base_install.sh" 0 1 (face ivy-virtual type file) 1 27 (face ivy-virtual)) #(".bin/finddups" 0 1 (face ivy-virtual type file) 1 13 (face ivy-virtual)) #(".bin/git-fix-remotes" 0 1 (face ivy-virtual type file) 1 20 (face ivy-virtual)) #(".bin/hashsync" 0 1 (face ivy-virtual type file) 1 13 (face ivy-virtual)) #(".bin/licblock" 0 1 (face ivy-virtual type file) 1 13 (face ivy-virtual)) #(".bin/netrc_pass" 0 1 (face ivy-virtual type file) 1 15 (face ivy-virtual)) #(".bin/pacman-rmorphans" 0 1 (face ivy-virtual type file) 1 21 (face ivy-virtual)) #(".bin/symlink_dotfiles.sh" 0 1 (face ivy-virtual type file) 1 24 (face ivy-virtual)) #(".bin/thblt_got_a_new_pc" 0 1 (face ivy-virtual type file) 1 23 (face ivy-virtual)) #(".bin/truth" 0 1 (face ivy-virtual type file) 1 10 (face ivy-virtual)) #(".bin/undocleaner" 0 1 (face ivy-virtual type file) 1 16 (face ivy-virtual)) #(".bin/wmake" 0 1 (face ivy-virtual type file) 1 10 (face ivy-virtual)) #(".bin/yaourt-install" 0 1 (face ivy-virtual type file) 1 19 (face ivy-virtual)) #(".config/dunst/dunstrc" 0 1 (face ivy-virtual type file) 1 21 (face ivy-virtual)) #(".config/gtk-3.0/gtk.css" 0 1 (face ivy-virtual type file) 1 23 (face ivy-virtual)) #(".config/gtk-3.0/settings.ini" 0 1 (face ivy-virtual type file) 1 28 (face ivy-virtual)) #(".config/synapse/config.json" 0 1 (face ivy-virtual type file) 1 27 (face ivy-virtual)) #(".config/user-dirs.dirs" 0 1 (face ivy-virtual type file) 1 22 (face ivy-virtual)) #(".config/user-dirs.locale" 0 1 (face ivy-virtual type file) 1 24 (face ivy-virtual)) #(".emacs.d/init.el" 0 1 (face ivy-virtual type file) 1 16 (face ivy-virtual)) #(".emacs.d/install-docsets.el" 0 1 (face ivy-virtual type file) 1 27 (face ivy-virtual)) #(".emacs.d/snippets/org-mode/author" 0 1 (face ivy-virtual type file) 1 33 (face ivy-virtual)) #(".emacs.d/snippets/org-mode/beg" 0 1 (face ivy-virtual type file) 1 30 (face ivy-virtual)) #(".emacs.d/snippets/org-mode/begin" 0 1 (face ivy-virtual type file) 1 32 (face ivy-virtual)) #(".emacs.d/snippets/org-mode/bold" 0 1 (face ivy-virtual type file) 1 31 (face ivy-virtual)) #(".emacs.d/snippets/org-mode/cl" 0 1 (face ivy-virtual type file) 1 29 (face ivy-virtual)) #(".emacs.d/snippets/org-mode/desc" 0 1 (face ivy-virtual type file) 1 31 (face ivy-virtual)) #(".emacs.d/snippets/org-mode/dita" 0 1 (face ivy-virtual type file) 1 31 (face ivy-virtual)) #(".emacs.d/snippets/org-mode/docbook" 0 1 (face ivy-virtual type file) 1 34 (face ivy-virtual)) #(".emacs.d/snippets/org-mode/el" 0 1 (face ivy-virtual type file) 1 29 (face ivy-virtual)) #(".emacs.d/snippets/org-mode/email" 0 1 (face ivy-virtual type file) 1 32 (face ivy-virtual)) #(".emacs.d/snippets/org-mode/ex" 0 1 (face ivy-virtual type file) 1 29 (face ivy-virtual)) #(".emacs.d/snippets/org-mode/figure" 0 1 (face ivy-virtual type file) 1 33 (face ivy-virtual)) #(".emacs.d/snippets/org-mode/html" 0 1 (face ivy-virtual type file) 1 31 (face ivy-virtual)) #(".emacs.d/snippets/org-mode/inc" 0 1 (face ivy-virtual type file) 1 30 (face ivy-virtual)) #(".emacs.d/snippets/org-mode/keywords" 0 1 (face ivy-virtual type file) 1 35 (face ivy-virtual)) #(".emacs.d/snippets/org-mode/lang" 0 1 (face ivy-virtual type file) 1 31 (face ivy-virtual)) #(".emacs.d/snippets/org-mode/latex" 0 1 (face ivy-virtual type file) 1 32 (face ivy-virtual)) #(".emacs.d/snippets/org-mode/options" 0 1 (face ivy-virtual type file) 1 34 (face ivy-virtual)) #(".emacs.d/snippets/org-mode/owh" 0 1 (face ivy-virtual type file) 1 30 (face ivy-virtual)) ...) :matcher counsel-projectile--matcher :require-match t :keymap (keymap (27 keymap (32 . counsel-projectile-drop-to-switch-project))) :action counsel-projectile--action :caller counsel-projectile)
  counsel-projectile()
  projectile-switch-project-by-name("~/.dotfiles/" nil)
  #[(dir) "\302	\"\207" [dir arg projectile-switch-project-by-name] 3]("~/.dotfiles/")
  ivy-call()
  ivy-read("[-] Switch to project: " ("~/.dotfiles/" "~/Documents/Code/Playground/" "~/Documents/Code/Thb.lt/") :preselect nil :action #[(dir) "\302	\"\207" [dir arg projectile-switch-project-by-name] 3] :require-match t :caller counsel-projectile-switch-project)
  counsel-projectile-switch-project(nil)
  call-interactively(counsel-projectile-switch-project nil nil)
  command-execute(counsel-projectile-switch-project)
```

**Comment by nixmaniack** [comment]
I have faced this and had to revert `ivy` to older version to get `counsel-projectile-switch-project` working . Here's what has changed in `ivy` and very likely cause of the bug.

```diff
diff --git a/ivy-20161126.533/ivy.el b/ivy-20161129.516/ivy.el
index 9ba6db5..970e25b 100644
--- a/ivy-20161126.533/ivy.el
+++ b/ivy-20161129.516/ivy.el
@@ -993,8 +993,7 @@ Example use:
           (if (eq action 'identity)
               (funcall action x)
             (select-window (ivy--get-window ivy-last))
-            (prog1 (let ((default-directory (ivy-state-directory ivy-last)))
-                     (funcall action x))
+            (prog1 (funcall action x)
               (unless (or (eq ivy-exit 'done)
                           (equal (selected-window)
                                  (active-minibuffer-window))
@@ -1364,7 +1363,7 @@ a hash table.
 PREDICATE is applied to filter out the COLLECTION immediately.
 This argument is for `completing-read' compat.

-When REQUIRE-MATCH is non-nil, only memebers of COLLECTION can be
+When REQUIRE-MATCH is non-nil, only members of COLLECTION can be
 selected, i.e. custom text.

 If INITIAL-INPUT is not nil, then insert that input in the
```

**Comment by ericdanan** [comment]
Thank you both for detailed reporting. Indeed the reason is a recent commit in ivy.

It should work now with the latest ivy version, can you confirm?

**Comment by thblt** [comment]
Indeed, I fixed it by [pinning](https://github.com/jwiegley/use-package) ivy to melpa-stable, but that's not really a clean solution...

**Comment by thblt** [comment]
@ericdanan  Unpinning and testing right now :-)

**Comment by nixmaniack** [comment]
@ericdanan That fixed it. Thank you!

**Comment by thblt** [comment]
Sorry, doesn't work better for me, after a full update.  The steps described in m first post now lead to the error:

```
Debugger entered--Lisp error: (wrong-type-argument stringp #<window 3 on *scratch*>)
  counsel-projectile--find-file-action(#("VocabDecTree/build/CMakeCache.txt" 0 1 (type file face ivy-virtual) 1 33 (face ivy-virtual)) nil)
  counsel-projectile--action(#("VocabDecTree/build/CMakeCache.txt" 0 1 (type file face ivy-virtual) 1 33 (face ivy-virtual)))
  ivy-call()
  ivy-read("[Playground] Load buffer or file: " (#("VocabDecTree/#sort.c#" 0 1 (face ivy-virtual type file) 1 21 (face ivy-virtual)) #("VocabDecTree/build/CMakeCache.txt" 0 1 (face ivy-virtual type file) 1 33 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/3.6.2/CMakeCCompiler.cmake" 0 1 (face ivy-virtual type file) 1 56 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/3.6.2/CMakeCXXCompiler.cmake" 0 1 (face ivy-virtual type file) 1 58 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/3.6.2/CMakeDetermineCompilerABI_C.bin" 0 1 (face ivy-virtual type file) 1 67 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/3.6.2/CMakeDetermineCompilerABI_CXX.bin" 0 1 (face ivy-virtual type file) 1 69 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/3.6.2/CMakeSystem.cmake" 0 1 (face ivy-virtual type file) 1 53 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/3.6.2/CompilerIdC/CMakeCCompilerId.c" 0 1 (face ivy-virtual type file) 1 66 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/3.6.2/CompilerIdCXX/CMakeCXXCompilerId.cpp" 0 1 (face ivy-virtual type file) 1 72 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/3.7.0/CMakeCCompiler.cmake" 0 1 (face ivy-virtual type file) 1 56 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/3.7.0/CMakeCXXCompiler.cmake" 0 1 (face ivy-virtual type file) 1 58 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/3.7.0/CMakeDetermineCompilerABI_C.bin" 0 1 (face ivy-virtual type file) 1 67 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/3.7.0/CMakeDetermineCompilerABI_CXX.bin" 0 1 (face ivy-virtual type file) 1 69 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/3.7.0/CMakeSystem.cmake" 0 1 (face ivy-virtual type file) 1 53 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/3.7.0/CompilerIdC/CMakeCCompilerId.c" 0 1 (face ivy-virtual type file) 1 66 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/3.7.0/CompilerIdCXX/CMakeCXXCompilerId.cpp" 0 1 (face ivy-virtual type file) 1 72 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/CMakeDirectoryInformation.cmake" 0 1 (face ivy-virtual type file) 1 61 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/CMakeOutput.log" 0 1 (face ivy-virtual type file) 1 45 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/Makefile.cmake" 0 1 (face ivy-virtual type file) 1 44 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/Makefile2" 0 1 (face ivy-virtual type file) 1 39 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/TargetDirectories.txt" 0 1 (face ivy-virtual type file) 1 51 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/cmake.check_cache" 0 1 (face ivy-virtual type file) 1 47 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/feature_tests.bin" 0 1 (face ivy-virtual type file) 1 47 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/feature_tests.c" 0 1 (face ivy-virtual type file) 1 45 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/feature_tests.cxx" 0 1 (face ivy-virtual type file) 1 47 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/progress.marks" 0 1 (face ivy-virtual type file) 1 44 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/voc2dc.dir/C.includecache" 0 1 (face ivy-virtual type file) 1 55 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/voc2dc.dir/DependInfo.cmake" 0 1 (face ivy-virtual type file) 1 57 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/voc2dc.dir/build.make" 0 1 (face ivy-virtual type file) 1 51 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/voc2dc.dir/cmake_clean.cmake" 0 1 (face ivy-virtual type file) 1 58 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/voc2dc.dir/depend.internal" 0 1 (face ivy-virtual type file) 1 56 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/voc2dc.dir/depend.make" 0 1 (face ivy-virtual type file) 1 52 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/voc2dc.dir/flags.make" 0 1 (face ivy-virtual type file) 1 51 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/voc2dc.dir/link.txt" 0 1 (face ivy-virtual type file) 1 49 (face ivy-virtual)) #("VocabDecTree/build/CMakeFiles/voc2dc.dir/progress.make" 0 1 (face ivy-virtual type file) 1 54 (face ivy-virtual)) #("VocabDecTree/build/Makefile" 0 1 (face ivy-virtual type file) 1 27 (face ivy-virtual)) #("VocabDecTree/build/cmake_install.cmake" 0 1 (face ivy-virtual type file) 1 38 (face ivy-virtual)) #("VocabDecTree/build/compile_commands.json" 0 1 (face ivy-virtual type file) 1 40 (face ivy-virtual)) #("VocabDecTree/build/voc2dc" 0 1 (face ivy-virtual type file) 1 25 (face ivy-virtual)) #(".gitignore" 0 1 (face ivy-virtual type file) 1 10 (face ivy-virtual)) #("README.md" 0 1 (face ivy-virtual type file) 1 9 (face ivy-virtual)) #("Sudoku/Solver.lhs" 0 1 (face ivy-virtual type file) 1 17 (face ivy-virtual)) #("VocabDecTree/CMakeLists.txt" 0 1 (face ivy-virtual type file) 1 27 (face ivy-virtual)) #("VocabDecTree/README.org" 0 1 (face ivy-virtual type file) 1 23 (face ivy-virtual)) #("VocabDecTree/initial_compiler.c" 0 1 (face ivy-virtual type file) 1 31 (face ivy-virtual)) #("VocabDecTree/sort.c" 0 1 (face ivy-virtual type file) 1 19 (face ivy-virtual)) #("VocabDecTree/sort.h" 0 1 (face ivy-virtual type file) 1 19 (face ivy-virtual)) #("VocabDecTree/vector.c" 0 1 (face ivy-virtual type file) 1 21 (face ivy-virtual)) #("VocabDecTree/vector.h" 0 1 (face ivy-virtual type file) 1 21 (face ivy-virtual))) :matcher counsel-projectile--matcher :require-match t :keymap (keymap (27 keymap (32 . counsel-projectile-drop-to-switch-project))) :action counsel-projectile--action :caller counsel-projectile)
  counsel-projectile()
  projectile-switch-project-by-name("~/Documents/Code/Playground/" nil)
  #[(dir) "\302	\"\207" [dir arg projectile-switch-project-by-name] 3]("~/Documents/Code/Playground/")
  ivy-call()
  ivy-read("[.dotfiles] Switch to project: " ("~/.dotfiles/" "~/Documents/Code/Playground/" "~/Documents/Code/Thb.lt/") :preselect "~/.dotfiles/" :action #[(dir) "\302	\"\207" [dir arg projectile-switch-project-by-name] 3] :require-match t :caller counsel-projectile-switch-project)
  counsel-projectile-switch-project(nil)
  call-interactively(counsel-projectile-switch-project nil nil)
  command-execute(counsel-projectile-switch-project)
```

**Comment by thblt** [comment]
In fact, Swiper seems broken as well.  I really don't have time to investigate this any further right now, so I'm pinning Ivy back to stable.  If you can't reproduce, I'll try to have a deeper look tomorrow.

**Comment by ericdanan** [comment]
Indeed I can't reproduce this. The steps from your first post work fine in `emacs -Q` for me, and so does swiper...

**Comment by thblt** [comment]
I now suspect my update led to a mix of melpa and melpa stable versions of Ivy and Swiper which led to weird behavior.  It may very well be my mistake.  Since it works for @nixmaniack, I suggest we close this.  Thanks a lot for the quick fix :-)

**Comment by nixmaniack** [comment]
I think we can close this. I don't see issue which @thblt is facing, swiper works fine at my end.

**Comment by ericdanan** [comment]
OK, closing.

**Comment by ericdanan** [comment]
I reverted the last commit because the same change has been implemented globally in ivy (swiper commit 50bb6b3).

It shouldn't change anything, but let me know if you encounter any issue.
